File: FpWorkshop/FSharpPortion/Step04_Options.fs
﻿namespace FSharpPortion

module Step04_Options =

  let run () =
    // NOTE: Do this entire file before trying the C# one

    // An option is an F# type that represents the absense or presence of a value

    // Many find it useful to think of options as special lists that are only allowed
    // to have one element (e.g. `[42]`) or no elements (e.g. `[]`).

    // ^^ this intuition is useful because it shows that options basically have all the
    // same capabilities as lists!

    // Challenge 1 --------------------------------------
    // We can check if a value is present or absent
    let present = Some 42  // Kinda like [42]
    let absent = None      // Kinda like []

    ensure (Option.isSome present = __)
    ensure (Option.isSome absent = __)

    ensure (Option.isNone present = __)
    ensure (Option.isNone absent = __)

    // Challenge 2 --------------------------------------
    // If options are like lists, `Option.get` is like `List.head`
    // Also similarly, this is unsafe and should be avoided in production
    // (more to say about this later)
    ensure (Option.get __ = 42)

    // Challenge 3 --------------------------------------
    // We can transform values inside of options using `map` (sound familiar?)

    ensure (Option.map (fun v -> __) (Some 43) = Some 42)

    // Challenge 4 --------------------------------------
    // If we `map` over an empty list we get the empty list. What do you think
    // mapping over `None` yields?

    ensure (Option.map (fun v -> v + "!") None = __)

    // Challenge 5 --------------------------------------
    // Unfortunately (or fortunately, depending on how you look at it) options' version
    // of `List.collect` is called `Option.bind` but it does the same thing! Map then squish.

    // I find it fun to think of how the "squish stage" might work so let's work with that first

    ensure (pretendSquish (Some None) = __)
    ensure (pretendSquish None = __)
    ensure (pretendSquish (Some (Some 99)) = __)

    // Challenge 6 --------------------------------------
    // So with actual `bind`...

    ensure (Option.bind (fun _ -> None)    None      = __)
    ensure (Option.bind (fun _ -> Some 42) None      = __)
    ensure (Option.bind (fun _ -> None)    (Some 42) = __)
    ensure (Option.bind (fun _ -> Some 99) (Some 42) = __)

    // SPOILER: any None makes the whole thing none!

    // Challenge 7 --------------------------------------
    // So what does that mean? Basically it means that we can write code
    // in the context that values might not be there without actually ever
    // talking about their absense!

    let result =
      Some 42 |> Option.bind (fun v1 ->
      None |> Option.bind (fun v2 ->
      Some 99 |> Option.map (fun v3 ->
        (v1 + v2) * v3
      )))

    ensure (result = __)

    // Challenge 8 --------------------------------------
    // Why is there a `map` in the innermost part above?
    // NOTE: this is an extremely common pattern in FP, `bind, bind, bind ... bind, map`

    // Challenge 9 --------------------------------------
    // In lists, remember how `head` is unsafe? Well, option to the rescue!
    // `tryHead` returns an option so it's a safe alternative.

    ensure (List.tryHead [] = __)
    ensure (List.tryHead [42; 3; 9] = __)

    // Challenge 10 --------------------------------------
    // If you've been skeptical about options being similar to lists this whole time
    // Here's a little evidence!

    ensure (Option.toList None = __)
    ensure (Option.toList (Some 42) = __)

    // How'd you get in here, array!?
    ensure (Option.toArray (Some 3) = __)
    ensure (Option.toArray None = __)

    // Challenge 11 --------------------------------------
    // If options are so simlar to lists then why not just use lists everywhere?

    // Challenge 12 --------------------------------------
    // Option has safer ways to get the value, like providing a default for when it's
    // not found

    ensure (Option.defaultValue 42 None = __)
    ensure (Option.defaultValue 42 (Some 99) = __)

File: FpWorkshop/FpWorkshop/Step04_Options.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;

namespace FpWorkshop
{
  public class Step04_Options
  {
    public interface IOption<T>
    {
      bool IsSome { get; }
      bool IsNone { get; }
      bool IsSomeOf(T value);
      IOption<TResult> Bind<TResult>(Func<T, IOption<TResult>> f);
      IOption<TResult> Map<TResult>(Func<T, TResult> f);
      T DefaultValue(T aDefault);
      IList<T> ToList();
    }

    public class None<T> : IOption<T>
    {
      public bool IsSome => false;
      public bool IsSomeOf(T value) => false;

      public bool IsNone => true;

      public IOption<TResult> Bind<TResult>(Func<T, IOption<TResult>> f) => new None<TResult>();

      public IOption<TResult> Map<TResult>(Func<T, TResult> f) => new None<TResult>();

      public T DefaultValue(T aDefault) => aDefault;

      public IList<T> ToList() => new List<T>();

      public override bool Equals(object obj) => obj is None<T>;
    }

    public class Some<T> : IOption<T>
    {
      private readonly T _v;

      public Some(T v)
      {
        _v = v;
      }

      public bool IsSomeOf(T value) => _v.Equals(value);

      public bool IsSome => true;

      public bool IsNone => false;

      public IOption<TResult> Bind<TResult>(Func<T, IOption<TResult>> f) =>
        f(_v);

      public IOption<TResult> Map<TResult>(Func<T, TResult> f) =>
        new Some<TResult>(f(_v));

      public T DefaultValue(T aDefault) => _v;

      public IList<T> ToList() => new List<T> {_v};
    }

    public static void Run()
    {
      // Challenge: Make all these tests pass by implementing option
      // (hehe :trollface:)

      // Extra Credit: make IOption implement IEnumerable (it's not too hard)

      // Challenge 1 --------------------------------------
      var present = new Some<int>(42);
      var absent = new None<int>();

      Help.Ensure(present.IsSome);
      Help.Ensure(!absent.IsSome);

      Help.Ensure(!present.IsNone);
      Help.Ensure(absent.IsNone);

      // Challenge 2 --------------------------------------
      // Let's ignore this in our implementation

      // Challenge 3 --------------------------------------
      Help.Ensure(new Some<int>(43).Map(v => v - 1).IsSomeOf(42));

      // Challenge 4 --------------------------------------
      Help.Ensure(new None<string>().Map(v => v + "!").IsNone);

      // Challenge 5 --------------------------------------
      // If bind is correct, squishing works :)

      // Challenge 6 --------------------------------------
      // So with actual `bind`...

      Help.Ensure(new None<int>().Bind(_ => new None<int>()).IsNone);
      Help.Ensure(new Some<int>(42).Bind(_ => new None<int>()).IsNone);
      Help.Ensure(new None<int>().Bind(_ => new Some<int>(42)).IsNone);
      Help.Ensure(new Some<int>(42).Bind(_ => new Some<int>(99)).IsSomeOf(99));

      // Challenge 7 --------------------------------------

      var result =
        new Some<int>(42).Bind(v1 =>
        new None<int>().Bind(v2 =>
        new Some<int>(99).Map(v3 =>
          (v1 + v2) * v3
        )));

      Help.Ensure(result.IsNone);

      var result2 =
        new Some<int>(42).Bind(v1 =>
        new Some<int>(2).Bind(v2 =>
        new Some<int>(99).Map(v3 =>
          (v1 + v2) * v3
        )));

      Help.Ensure(result2.IsSomeOf((42 + 2) * 99));

      // Challenge 8 --------------------------------------

      // Challenge 9 --------------------------------------

      // Challenge 10 --------------------------------------

      Help.Ensure(!new None<int>().ToList().Any());
      Help.Ensure(new Some<int>(42).ToList().SequenceEqual(new List<int>{ 42 }));

      // Challenge 11 --------------------------------------
      // If options are so simlar to lists then why not just use lists everywhere?

      // Challenge 12 --------------------------------------
      // Option has safer ways to get the value, like providing a default for when it's
      // not found

      Help.Ensure(new None<int>().DefaultValue(42) == 42);
      Help.Ensure(new Some<int>(99).DefaultValue(42) == 99);
    }
  }
}